### PR TITLE
Add the ability to create custom vertex formats

### PR DESCRIPTION
**Pull request by walterpie**
This commit makes it possible to create vertex formats
differing from the default position/normal/uv vertex.
It doesn't change much functionality and uses existing
APIs.

NOTE: This might not be the most optimal approach. 
It achieves what it achieves by compiling and specializing
pipelines BEFORE the mesh resource provider system.
So please make suggestions how I can make this better.

### CODE REVIEW

**Comment by walterpie** [comment]
PS: I did this because I'm working on a voxel renderer for bevy.

**Comment by StarArawn** [comment]
Just a heads up there is an existing draft PR open for this as well:
https://github.com/bevyengine/bevy/pull/158

File: crates/bevy_render/src/mesh/mesh.rs [hunk 1]
```diff
@@ -481,21 +510,14 @@ pub fn mesh_resource_provider_system(
...
-            let vertex_buffer_descriptor = Vertex::as_vertex_buffer_descriptor();
-            vertex_buffer_descriptors.set(vertex_buffer_descriptor.clone());
-            state.vertex_buffer_descriptor = Some(vertex_buffer_descriptor);
-            vertex_buffer_descriptor
-        }
-    };
     let mut changed_meshes = HashSet::new();
-    let render_resource_context = &**render_resource_context;
     for event in state.mesh_event_reader.iter(&mesh_events) {
+        let render_resource_context = &**render_resource_context;
```
Comment (0xWind) [inline crates/bevy_render/src/mesh/mesh.rs:517]: Why move here?

**Comment by walterpie** [comment]
Thanks for the tip! Maybe I can contribute to their solution. I had the same issues as they had, but I managed to get a working solution.

**Comment by bonsairobo** [comment]
> PS: I did this because I'm working on a voxel renderer for bevy.

Hey @walterpie let me know if you want to team up on this. I'm planning to do the same thing. I've made some progress on a voxel renderer, etc for the Amethyst engine, and I want to port it to Bevy: https://github.com/amethyst/voxel-mapper

**Comment by walterpie** [comment]
Hey @bonsairobo, I just added you. Let's discuss the details over on Discord.

**Comment by cart** [comment]
Closing in favor of the direction we're taking in #599 (and future follow up prs)

**Pull request closed without merging**